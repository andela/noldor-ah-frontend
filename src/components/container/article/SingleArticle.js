/* eslint-disable max-len */
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import SingleArticle from '../../presentational/article/singleArticle';
import { loadAnArticle } from '../../../redux/actions/single-article/singleArticleAction';
import CategoryList from '../../presentational/category-list/CategoryList';
import Loading from '../../common/loader/Loader';

import './style.scss';

export class SinglePage extends Component {
    componentDidMount = () => {
      // window.scrollTo(0, 0);
      const { slug } = this.props;
      this.props.singleArticle(slug);
    }

    render() {
      const { anArticle } = this.props;
      if (anArticle.isLoading !== false) {
        return <Loading />;
      }

      return (

        <section >

          <CategoryList />
          <SingleArticle article={anArticle.article} />
        </section>
      );
    }
}

SinglePage.propTypes = {
  history: PropTypes.object,
  slug: PropTypes.string,
  anArticle: PropTypes.object.isRequired,
  singleArticle: PropTypes.func.isRequired

};


export const mapStateToProps = (state, ownProp) => {
  // console.log(state)
  return {
    slug: ownProp.match.params.slug,
    anArticle: state.anArticle
  };
};

export const mapDispatchToProps = {
  singleArticle: slug => loadAnArticle(slug)
};

export default connect(mapStateToProps, mapDispatchToProps)(SinglePage);
