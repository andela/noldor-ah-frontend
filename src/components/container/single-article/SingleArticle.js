/* eslint-disable max-len */
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { ToastContainer, notifier } from '../../../utilities/toast/notifier';
import SingleArticle from '../../presentational/article/singleArticle';
import { loadAnArticle } from '../../../redux/actions/single-article/singleArticleAction';
import CategoryList from '../../common/category-list/CategoryList';
import '../styles/SingleArticle.scss';
import Loading from '../../common/loader/Loader';

import '../../presentational/styles/comment.scss';
import { rateArticles } from '../../../redux/actions/article-rating/articleRating';
import { getUserRating } from '../../../redux/actions/article-rating/userArticleRating';

export class SinglePage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      rate: {}
    };
    this.rate = 0;
    this.articleId = '';
    this.handleOnChangeEvent = this.handleOnChangeEvent.bind(this);
  }

  componentDidMount = () => {
    const { slug } = this.props;
    this.props.singleArticle(slug);
  }

  async handleOnChangeEvent(rateValue) {
    const id = this.articleId;
    const data = {
      id,
      rateValue
    };
    const response = await this.props.rateAnArticle(data);
    if (response.type === 'RATE_ARTICLE_SUCCESS') {
      notifier('Article rated succesfully', 'success');
    } else {
      notifier(response.error.data.message, 'failure');
    }
  }


  render() {
    const { anArticle } = this.props;
    const { ratings } = anArticle.article;
    this.articleId = anArticle.article.id;

    if (this.articleId) {
      this.props.userRate(this.articleId);
    }

    if (anArticle.isLoading !== false) {
      return <Loading />;
    }
    return (
      <div>
        <ToastContainer />
        <section >
          <CategoryList />
          <SingleArticle article={anArticle.article}
            eventHandler={this.handleOnChangeEvent}
            averageRate={ratings}
            userRating={this.rate}
          />
        </section>
      </div>

    );
  }
}

SinglePage.propTypes = {
  history: PropTypes.object,
  slug: PropTypes.string,
  anArticle: PropTypes.object.isRequired,
  singleArticle: PropTypes.func.isRequired,
  rateAnArticle: PropTypes.func,
  userRate: PropTypes.func
};


export const mapStateToProps = (state, ownProp) => {
  return {
    slug: ownProp.match.params.slug,
    anArticle: state.anArticle,
    rateArticle: state.articleRating,
    userRates: state.userRatings
  };
};

export const mapDispatchToProps = {
  singleArticle: slug => loadAnArticle(slug),
  rateAnArticle: data => rateArticles(data),
  userRate: articleId => getUserRating(articleId)
};

export default connect(mapStateToProps, mapDispatchToProps)(SinglePage);
