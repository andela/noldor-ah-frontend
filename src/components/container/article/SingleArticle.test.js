import React from 'react';
import { mount, shallow } from 'enzyme';
import { SinglePage, mapDispatchToProps, mapStateToProps } from './SingleArticle';
// import { loadAnArticle } from '../../../actions/singlearticle/singleArticleAction';

describe('landing page', () => {
  const getProps = isLoading => (
    {
      anArticle: {
        isLoading,
        article: []
      },
      relatedArticle: {
        isLoading,
        articles: []
      },
      allArticles: [],
      singleArticle: () => {},
      relateArticle: () => {},
      allArticle: () => {}
    }
  );
  it('should show page is loading', () => {
    const props = getProps(true);
    const component = shallow(<SinglePage {...props}/>);
    expect(component).toMatchSnapshot();
  });

  it('should show page is loading', () => {
    const props = getProps(false);
    const component = shallow(<SinglePage {...props}/>);
    expect(component).toMatchSnapshot();
  });

  it('should show loading if props havent returned', () => {
    const props = getProps(true);
    const wrapper = mount(<SinglePage {...props} />);
    expect(wrapper.find('Loading')).toBeDefined();
    expect(wrapper.find('Loader')).toBeDefined();
    expect(wrapper.find('Hearts')).toBeDefined();
    expect(wrapper.find('svg')).toHaveLength(1);
  });

  it('should show page after loading ', () => {
    const props = getProps(false);
    const wrapper = mount(<SinglePage {...props} />);
    expect(wrapper.find('footer-article')).toBeDefined();
  });

  it('define mapStateToProps', () => {
    const ownProp = {
      match: {
        params: {
          slug: ''
        }
      }
    };
    const state = {};
    expect(mapStateToProps(state, ownProp)).toBeDefined();
  });
  it('define mapDispatchToProps', () => {
    // const wrapper = shallow(<Connect  />)
    // console.log(wrapper.debug())
    // const props = {
    //   singleArticle: () => loadAnArticle(),

    // };

    expect(mapDispatchToProps).toBeDefined();
    // expect(mapDispatchToProps).toHaveReturned({ ...props });
  });
});
