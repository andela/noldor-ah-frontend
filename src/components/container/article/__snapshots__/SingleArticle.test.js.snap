// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`landing page should show page is loading 1`] = `<Loading />`;

exports[`landing page should show page is loading 2`] = `
<section>
  <Connect(CategoriesList) />
  <div
    className="base-line theme-background"
  />
  <SingleArticle
    article={Array []}
  />
</section>
`;
